### PR TITLE
Add callbacks via markers

### PR DESCRIPTION
**Pull request by Pezzah**
## Goal

Allow meta data to be added to individual logger statements via a custom marker object

## Design

Previously we used a custom exception, but using a marker is viewed as a slightly cleaner approach

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: bugsnag/src/main/java/com/bugsnag/logback/BugsnagMarker.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return BUGSNAG_MARKER_NAME;
+    }
+
+    @Override
+    public void add(Marker reference) {
+
```
Comment (kattrali) [inline bugsnag/src/main/java/com/bugsnag/logback/BugsnagMarker.java]: Shouldn't there be a `references.add` here?

Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/logback/BugsnagMarker.java]: added tests

File: bugsnag/src/main/java/com/bugsnag/BugsnagAppender.java [hunk 2]
```diff
@@ -140,6 +141,13 @@ protected void append(final ILoggingEvent event) {
         if (bugsnag != null) {
             Throwable throwable = extractThrowable(event);
 
+            final Callback reportCallback;
+            if (event.getMarker() != null && event.getMarker() instanceof BugsnagMarker) {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/BugsnagAppender.java]: changed

File: bugsnag/src/main/java/com/bugsnag/logback/BugsnagMarker.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    public BugsnagMarker(Callback callback) {
+        this.callback = callback;
+    }
+
+    public Callback getCallback() {
+        return callback;
+    }
+
+    public void setCallback(Callback callback) {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/logback/BugsnagMarker.java]: We can probably remove this method

File: bugsnag/src/main/java/com/bugsnag/logback/BugsnagMarker.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import org.slf4j.Marker;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Marker used to pass a callback through a logger call
+ */
+public class BugsnagMarker implements Marker {
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply bugsnag/src/main/java/com/bugsnag/logback/BugsnagMarker.java]: added
